src/protocol/vm/state_builder.rs

```diff
@@ -0,0 +1,461 @@
+    involved_contracts: Option<HashSet<H160>>,
+    stateless_contracts: Option<HashMap<String, Option<Vec<u8>>>>,
+    token_storage_slots: Option<HashMap<H160, (ERC20Slots, ContractCompiler)>>,
+    manual_updates: Option<bool>,
+    trace: Option<bool>,
+    engine: Option<SimulationEngine<PreCachedDB>>,
+    adapter_contract: Option<TychoSimulationContract<PreCachedDB>>,
+    adapter_contract_path: Option<String>,
+}
+
+impl VMPoolStateBuilder {
+    pub fn new(id: String, tokens: Vec<H160>, block: BlockHeader) -> Self {
+        Self {
+            id,
+            tokens,
+            block,
+            balances: None,
+            balance_owner: None,
+            capabilities: None,
+            involved_contracts: None,
+            stateless_contracts: None,
+            token_storage_slots: None,
+            manual_updates: None,
+            trace: None,
+            engine: None,
+            adapter_contract: None,
+            adapter_contract_path: None,
+        }
+    }
+
+    pub fn balances(mut self, balances: HashMap<H160, U256>) -> Self {
+        self.balances = Some(balances);
+        self
+    }
+
+    pub fn balance_owner(mut self, balance_owner: H160) -> Self {
+        self.balance_owner = Some(balance_owner);
+        self
+    }
+
+    pub fn capabilities(mut self, capabilities: HashSet<Capability>) -> Self {
+        self.capabilities = Some(capabilities);
+        self
+    }
+
+    pub fn involved_contracts(mut self, involved_contracts: HashSet<H160>) -> Self {
+        self.involved_contracts = Some(involved_contracts);
+        self
+    }
+
+    pub fn stateless_contracts(
+        mut self,
+        stateless_contracts: HashMap<String, Option<Vec<u8>>>,
+    ) -> Self {
+        self.stateless_contracts = Some(stateless_contracts);
+        self
+    }
+
+    pub fn token_storage_slots(
+        mut self,
+        token_storage_slots: HashMap<H160, (ERC20Slots, ContractCompiler)>,
+    ) -> Self {
+        self.token_storage_slots = Some(token_storage_slots);
+        self
+    }
+
+    pub fn manual_updates(mut self, manual_updates: bool) -> Self {
+        self.manual_updates = Some(manual_updates);
+        self
+    }
+
+    pub fn trace(mut self, trace: bool) -> Self {
+        self.trace = Some(trace);
+        self
+    }
+
+    pub fn engine(mut self, engine: SimulationEngine<PreCachedDB>) -> Self {
+        self.engine = Some(engine);
+        self
+    }
+
+    pub fn adapter_contract(
+        mut self,
+        adapter_contract: TychoSimulationContract<PreCachedDB>,
+    ) -> Self {
+        self.adapter_contract = Some(adapter_contract);
+        self
+    }
+
+    pub fn adapter_contract_path(mut self, adapter_contract_path: String) -> Self {
+        self.adapter_contract_path = Some(adapter_contract_path);
+        self
+    }
+
+    /// Build the final VMPoolState object
+    pub async fn build(mut self) -> Result<VMPoolState<PreCachedDB>, SimulationError> {
```

Comment:
I think this is pretty good. I'd just add more comments describing the steps to create & setup an engine, as it might be not clear for someone who is not familiar with this code.

Reply:
Addressed in https://github.com/propeller-heads/tycho-simulation/pull/63/commits/07651ef573969400157bc6be46b0f7057f50b574